rfcs/20200710-reduction/README.md

```diff
@@ -0,0 +1,159 @@
+frameworks:
+
+| Algotirhm | Formula                      | TensorFlow | PyTorch | OpenVino |
+| ---------:| ----------------------------:| ----------:| -------:| --------:|
+| And       | and(x[:])                    |            |         | x        |
+| Or        | or(x[:])                     |            |         | x        |
+| Min       | min(x[:])                    | x          | x       | x        |
+| Max       | max(x[:])                    | x          | x       | x        |
+| Sum       | sum(x[:])                    | x          | x       | x        |
+| Prod      | mul(x[:])                    | x          | x       | x        |
+| Mean      | sum(x[:]) / n                | x          | x       | x        |
+| Variance  | Mean(pow(x[:] - Mean(x), 2)) | x          | x       |          |
+| Std       | pow(Variance(x), 0.5)        | x          | x       |          |
+| LogSumExp | log(sum(exp(x[:] - Max(x)))) + Max(x) | x | x       |          |
+| Median    |                              |            | x       |          |
+| Mode      |                              |            | x       |          |
+| p-norm    | pow(sum(pow(abs(x[:]), p)), 1/p) | x      | x       |          |
+
+The following algorithms are out of scope due to low priority:
+* And;
+* Or;
+* Median;
+* Mode.
+
+Also there are normalize operations, which variate between frameworks:
+* OpenVino:
+  L2_normalize(x) = x[:] / pow(eps_mode(sum(pow(x[:], 2)), eps), 0.5)
+    = x[:] / pow(eps_mode((pow(norm_l2(x[:]), 2)), eps), 0.5),
+  where eps_mode = {add, max}
+* TensorFlow:
+  L2_normalize(x) = x[:] / pow(max(sum(pow(x[:], 2)), eps), 0.5)
+    = x[:] / pow(max((pow(norm_l2(x[:]), 2)), eps), 0.5)
+* PyTorch:
+  L2_normalize(x) = x[:] / max(pow(sum(pow(x[:], 2)), 0.5), eps)
+    = x[:] / max(norm_l2(x[:]), eps)
+
+# Proposal
+## Option 1: Implement reduction primitive and norm primitive.
+
+### Motivation
+Motivation for this option is to keep reduction primitive API simple, but limit
+flexibility and amount of supported algorithms.
+
+### API
+
+``` cpp
+// dnnl_types.h
+typedef enum {
+    ...
+    dnnl_eltwise_pow,
+    ...
+    dnnl_reduction_max,
+    dnnl_reduction_min,
+    dnnl_reduction_norm_l1,
+    dnnl_reduction_norm_l2,
+    dnnl_reduction_norm_linf,
+    dnnl_reduction_sum,
+    dnnl_reduction_mul,
+    dnnl_reduction_mean,
+    ...
+} dnnl_alg_kind_t;
+
+typedef struct {
+    dnnl_primitive_kind_t primitive_kind;
+    // The kind of reduction algorithm.
+    dnnl_alg_kind_t reduction_alg_kind;
+    // Source memory descriptor.
+    dnnl_memory_desc_t src_desc;
+    // Destination memory descriptor.
+    dnnl_memory_desc_t dst_desc;
+} dnnl_reduction_desc_t;
+
+// dnnl.h
+dnnl_status_t DNNL_API dnnl_reduction_desc_init(
+        dnnl_reduction_desc_t *reduction_desc,
+        dnnl_alg_kind_t reduction_alg_kind,
+        const dnnl_memory_desc_t *src_desc,
+        const dnnl_memory_desc_t *dst_desc);
```

Comment:
Could you please say a few words on the relation between `dst_desc` and `src_desc`?

Reply:
specifically:
- broadcast (well, reduction) semantics support
- data types relationship
- memory format relationship

Reply:
dst_desc is required:
1. To choose reduction dimensions: For example: src_desc.dims = {3, 5} , dst_desc.dims = {3, 1}
2. To choose dst data type: For example, src_desc.data_type = bf16, dst_data_type = {bf16, f32}

reduction primitive doesn't change memory format. User provides memory descriptor with format_kind = any, so reduction will fill all necessary fields based on src format_kind.

P. S. Initially i was thinking about mask similar to the one used in output scale, but since user also might want to choose dst data type i decided to add a descriptor.

Reply:
Thanks! LGTM.